mdsa/algorithms/nystrom.py

```diff
@@ -0,0 +1,405 @@
+        """
+
+        if not seed_distmat.shape[0] < seed_distmat.shape[1]:
+            raise ValueError("seed distance matrix should"
+                             "have less rows than column")
+        if not dimensions <= seed_distmat.shape[0]:
+            raise ValueError("number of rows of seed matrix must be "
+                             ">= requested dim")
+
+        nseeds = seed_distmat.shape[0]
+        nfull = seed_distmat.shape[1]
+
+        # matrix E: extract columns 1--nseed
+        #
+        matrix_e = seed_distmat[:, 0:nseeds]
+        # print("INFO: Extracted Matrix E which is of shape %dx%d" %
+        #      (matrix_e.shape))
+
+        # matrix F: extract columns nseed+1--end
+        #
+        matrix_f = seed_distmat[:, nseeds:]
+        # print("INFO: Extracted Matrix F which is of shape %dx%d" %
+        #      (matrix_f.shape))
+
+        # matrix A
+        #
+        # print("INFO: Computing Matrix A")
+        t0 = time.clock()
+        matrix_a = self._calc_matrix_a(matrix_e)
+        if PRINT_TIMINGS:
+            print("TIMING(%s): Computation of A took %f CPU secs" %
+                  (__name__, time.clock() - t0))
+
+        # matrix B
+        #
+        # print("INFO: Computing Matrix B")
+        t0 = time.clock()
+        matrix_b = self._calc_matrix_b(matrix_e, matrix_f)
+        if PRINT_TIMINGS:
+            print("TIMING(%s): Computation of B took %f CPU secs" %
+                  (__name__, time.clock() - t0))
+
+        # print("INFO: Eigendecomposing A")
+        t0 = time.clock()
+        # eigh: eigen decomposition for symmetric matrices
+        # returns: w, v
+        # w : ndarray, shape (M,)
+        #     The eigenvalues, not necessarily ordered.
+        # v : ndarray, or matrix object if a is, shape (M, M)
+        #     The column v[:, i] is the normalized eigenvector corresponding
+        #     to the eigenvalue w[i].
+        # alternative is svd: [U, S, V] = numpy.linalg.svd(matrix_a)
+        (eigval_a, eigvec_a) = eigh(matrix_a)
+        if PRINT_TIMINGS:
+            print("TIMING(%s): Eigendecomposition of A took %f CPU secs" %
+                  (__name__, time.clock() - t0))
+        # Sort descending
+        ind = argsort(eigval_a)
+        ind = ind[::-1]
+        eigval_a = eigval_a[ind]
+        eigvec_a = eigvec_a[:, ind]
+
+        # print("INFO: Estimating MDS coords")
+        t0 = time.clock()
+        result = zeros((nfull, dimensions))  # X in Platt 2005
+        # Preventing negative eigenvalues by using abs value. Other option
+        # is to set negative values to zero. Fabian recommends using
+        # absolute values (as in SVD)
+        sqrt_eigval_a = sqrt(abs(eigval_a))
+        for i in xrange(nfull):
+            for j in xrange(dimensions):
+                if i + 1 <= nseeds:
+                    val = sqrt_eigval_a[j] * eigvec_a[i, j]
+                else:
+                    # - original, unoptimised code-block
```

Comment:
note, if optimization is a goal here, then it may be possible to further tweak this without a huge effort: it looks like the branching logic can be removed from the nested forloop by doing the compute for `range(0, min(nseeds, nfull) + 1)` separately. in which case, I think it can also be done in bulk:

```python
result[:nseeds+1, :] = sqrt_eigval_a * eigvec_a[:nseeds+1, :]
```

...although, totally untested...